matrix valued GFs with efficient storage

kleinhenz opened the pull request:
This implements

* matrix valued Green's functions (scalar case is still available via type parameter)
* efficient storage for GF objects which takes into account time symmetries
* interpolation for GF objects
* some misc cleanup

HugoStrand committed:
matrix valued time_gf

HugoStrand committed:
fix view call

HugoStrand committed:
[time_gf] scalar and matrix TimeGF subtypes TimeScalarGF, TimeMatrixGF

kleinhenz committed:
[time_gf] work on matrix valued TimeGF

kleinhenz committed:
[time_gf] store norb as type parameter

kleinhenz committed:
[time_grid] add ridx field to TimeGridPoint

* ridx stores index into physical components of G indexed by real/imag times

kleinhenz committed:
[time_gf] improve performance in scalar case

kleinhenz committed:
[examples] fix nca example

kleinhenz committed:
[benchmark] add benchmark.jl

* adds benchmark for performance of TimeGF indexing

kleinhenz committed:
[benchmark] add raw integer indexing benchmark for comparison

kleinhenz committed:
[misc] explicitly extend Base functions

kleinhenz committed:
[time_gf] fix construction of TimeGF on two branch contour

kleinhenz committed:
[dos] add AbstractDOS supertype and DOS wrapper type

* useful to have abstract dos type for defining TimeGF constructors
  instead of relying on separate dos2gf function

kleinhenz committed:
[storage] add memory efficient container for anti-Hermitian matrices

kleinhenz committed:
[time_gf] minor refactor

* parameterize AbstractTimeGF on element type
* define methods for extracting GF components on abstract type

kleinhenz committed:
[time_gf] compute norb type parameter in inner constructor

kleinhenz committed:
[time_gf] add FullTimeGF

* uses compact storage
* takes time symmetries into account

kleinhenz committed:
[time_gf] add missing constructor

kleinhenz committed:
[benchmark] benchmark indexing performance for FullTimeGF

kleinhenz committed:
[time_gf] change type parameter from integer norb to bool scalar

* When storing a container of Green's functions it will often be
  the case that we statically know whether they are all
  scalar/non-scalar but not their specific sizes. This allows us
  to make that case type stable.

kleinhenz committed:
[time_gf] implement time invariant gf on full contour + dos constructors

kleinhenz committed:
[time_gf] implement green's functions on two branch contour

kleinhenz committed:
[dos] make SingularDOS subtype of AbstractDOS

kleinhenz committed:
[contour] add convenience getindex method for extracting a branch from a contour

kleinhenz committed:
[time_gf] fix dos constructor

kleinhenz committed:
[contour] fix heaviside function taking contour argument

kleinhenz committed:
[storage] remove assert when setting diagonal elements of anti herm storage

* this may be violated numerically so doesn't make sense for container to enforce it

kleinhenz committed:
[time_gf] wip on fermionic/bosonic GFs

kleinhenz committed:
[misc] move using statements to top level file

kleinhenz committed:
[time_gf] correctly apply xi factor for matsubara component

kleinhenz committed:
[misc] wip make separate types for different contours/timegrids

kleinhenz committed:
[time_gf] introduce GFSignEnum for keeping track of fermionic/bosonic GFs

kleinhenz committed:
[storage] ImaginaryTimeStorage -> PeriodicStorage

kleinhenz committed:
[time_gf] change order of GenericTimeGF type parameters

kleinhenz committed:
[benchmark] update benchmark

kleinhenz committed:
[time_gf] add inline annotation to getindex methods

kleinhenz committed:
[time_gf] add interpolation for GenericTimeGF

kleinhenz committed:
[contour] add function for getting branch points from contour

kleinhenz committed:
[benchmark] add preliminary interp benchmark

krivenko committed:
Fix some issues

* Fix a bug in `heaviside(c::AbstractContour, t1::BranchPoint, t2::BranchPoint)`.
* Use the numerically stable function `fermi()` in `flat_dos()`.
* Remove an unreachable return statement in `src/gf/generic.jl`.
* Fix a few comments ('akoi' -> 'Aoki').

kleinhenz committed:
[storage] rename some containers

* PeriodicStorage -> CirculantStorage
* TimeInvariantAntiHermitianStorage -> AntiHermitianToeplitzStorage

kleinhenz committed:
[time_grid] fix realtimes/imagetimes functions

kleinhenz committed:
[storage] allow specifying how to compute diagonal elements via greater parameter

kleinhenz committed:
[time_gf] fix calculation of lesser diagonal elements of imaginary component

kleinhenz committed:
[time_gf] directly use beta parameter of contour

kleinhenz committed:
[time_gf] add ImaginaryTimeGF

kleinhenz committed:
[time_gf] define interpolation for AbstractTimeGF

* defined in terms of element access which is specialized for each subtype
* TODO think about how to avoid allocations in matrix valued cased

kleinhenz committed:
[contour] call operator for contour calls get_point

kleinhenz committed:
[benchmark] remove reference results

* doesn't make much sense to have these checked in to repo
  since results will be machine dependent

kleinhenz committed:
[contour] remove unused

kleinhenz committed:
[dos] add DeltaDOS type + add GaussKronrodDOSIntegrator type

* DeltaDOS represents a weighted sum of delta functions.
  This is more convenient/general than the previous gf1level function.
* GaussKronrodDOSIntegrator stores parameters for quadgk. Introducing
  type makes parameters easier to modify + adds clearer interface
  for new integrators.

kleinhenz committed:
[time_gf] fix bug in matrix valued gf

* replace conj with adjoint

kleinhenz committed:
[hdf5] implement save/load of GF data in alpscore compatible format

kleinhenz committed:
[hdf5] implement loading of GF data from nessi hdf5 format

kleinhenz committed:
[misc] cleanup

kleinhenz committed:
[time_gf] add similar and zero methods for getting an unitialized GF of the same type

kleinhenz committed:
[time_gf] add dos constructor methods to GenericTimeGF

kleinhenz committed:
[misc] fix some exports

kleinhenz committed:
[nca] fix nca example

kleinhenz committed:
[storage] mark getindex method inline

kleinhenz committed:
[nca] add atol parameter

kleinhenz committed:
[nca] dyson -> dyson! since we modify arguments

kleinhenz committed:
[doc] update README

kleinhenz committed:
[test] remove unused

kleinhenz committed:
[dos] normalize DeltaDOS when constructing from list of energies

krivenko commented on README.md:
```diff
@@ -10,15 +10,4 @@ Additionally, it provides functions for generating Green's functions, performing
 Credit to Andrey Antipov and Igor Krivenko for designing a first version of the abstractions implemented here.
 
 ## Usage
-The following code generates a non-equilibrium Green's function from a spectral density and saves it to an hdf5 archive
```
Having an example in a README file is always nice.
I would not expect this example to be hard to port to the new API.

kleinhenz replied:
done

krivenko commented on src/alps.jl:
```diff
@@ -0,0 +1,153 @@
+  data::Array{ComplexF64}
+end
+
+function ALPSComplex(g::HDF5.Dataset)
```
Please, add a compat record for `HDF5>=v0.14.0` (cf. https://github.com/JuliaIO/HDF5.jl/pull/690).
`

krivenko commented on src/time_grid.jl:
```diff
@@ -89,7 +130,27 @@ function integrate(f, grid::TimeGrid, t1::TimeGridPoint, t2::TimeGridPoint)
+imagtimes(grid::FullTimeGrid) = range(0.0, grid.contour.β, length=grid.ntau)
+imagtimes(grid::ImaginaryTimeGrid) = range(0.0, grid.contour.β, length=grid.ntau)
+
+struct TimeDomain
```
What is the purpose of this structure? A comment here would certainly help.

krivenko commented on src/time_grid.jl:
```diff
@@ -89,7 +130,27 @@ function integrate(f, grid::TimeGrid, t1::TimeGridPoint, t2::TimeGridPoint)
+  else
+    npts = t.domain == imaginary_branch ? grid.ntau : grid.nt
+    step = 1 / (npts - 1)
+    delta = Int(floor(t.ref / step))
```
This line seems redundant.

krivenko commented on src/gf.jl:
```diff
@@ -0,0 +1,189 @@
+abstract type AbstractTimeGF{T, scalar} end
+#TODO decide on full interface that should be supported by subtypes
```
Is this comment still relevant?

kleinhenz replied:
I've added a comment with the current interface

krivenko commented on src/gf.jl:
```diff
@@ -0,0 +1,189 @@
+  elseif component == :matsubara
+    X = real(-im .* G[imaginary_branch, imaginary_branch])
+    return norb == 1 ? X[:,1] : X[:,:,:,1]
+  elseif component == :retarded
```
I wonder if it would make sense to follow $G^{ret}$'s definition more closely and multiply the difference by a step function of the physical times (same for the advanced component).

kleinhenz replied:
going to leave this for later

krivenko commented on src/gf.jl:
```diff
@@ -0,0 +1,189 @@
+  end
+end
+
+function interpolate!(x, G::AbstractTimeGF{T, false}, t1::BranchPoint, t2::BranchPoint) where T
```
What is `x` here and how is this function different from `interpolate()`? What does it return? Also, why is it defined only for the `scalar = false` case?

kleinhenz replied:
The result of the interpolation is written into x so you don't have to allocate the result every time if you don't want to. There's not much point in doing this for the scalar case.

krivenko commented on src/storage/generic.jl:
```diff
@@ -0,0 +1,73 @@
+  end
+end
+
+function Base.setindex!(X::GenericStorage{T,true}, v, i, j) where T
```
Isn't this case covered by the previous method definition?

krivenko commented on src/storage/generic.jl:
```diff
@@ -0,0 +1,73 @@
+  T(zero(X.data))
+end
+
+function Base.:+(X::GenericStorage{T,scalar}, Y::GenericStorage{T,scalar}) where {T,scalar}
```
Is it possible to define these operations once for `AbstractStorage{T, scalar}` and without loss of performance?
Something like
```julia
function Base.:+(X::Storage, Y::Storage) where {Storage <: AbstractStorage{T, scalar}} where {T,scalar}
  @assert size(X) == size(Y)
  return Storage(X.data .+ Y.data, scalar)
end
```

kleinhenz replied:
done

krivenko commented on src/time_grid.jl:
```diff
@@ -1,79 +1,120 @@
-  idx::Int64
+  idx::Int64 # contour index
+  ridx::Int64 # real/imag time index
   val::BranchPoint
```
How about we move away from the inherited naming convention and give this field a more intelligible name, `bpoint` or something similar? It is quite a pain to see expressions like `t1.val.val`.

kleinhenz replied:
good idea! done.

krivenko commented on src/gf/keldysh.jl:
```diff
@@ -0,0 +1,92 @@
+  grid::KeldyshTimeGrid
+  gtr::AntiHermitianStorage{T,scalar}
+  les::AntiHermitianStorage{T,scalar}
+  ξ::GFSignEnum
```
Is `ξ` actually used by `KeldyshTimeGF`?

kleinhenz replied:
Yeah this is not actually used in the `KeldyshTimeGF` code, but I left it since I thought you might want to keep track of if the GF is fermionic or bosonic.

krivenko commented on src/gf/keldysh.jl:
```diff
@@ -0,0 +1,92 @@
+
+  for i in 1:nt
+    for j in 1:i
+      push!(points, (tminus[j], tminus[i])) # greater
```
Perhaps I miss something here, but why do you have `tminus` twice?

kleinhenz replied:
This is the lower triangular portion of the (--) block which is where we define the greater GF.

krivenko commented on src/gf/full.jl:
```diff
@@ -0,0 +1,135 @@
+  j = t2.ridx
+  ξ = Int(G.ξ)
+
+  if ((t1.val.domain == forward_branch || t1.val.domain == backward_branch) &&
```
Here is a simple way to slightly improve clarity of the code below,
```julia
real_branches = (forward_branch, backward_branch)
if ((t1.val.domain in real_branches) && ...
```
You can also consider making `real_branches` a global constant defined in `src/branch.jl`.

kleinhenz replied:
done

krivenko commented on src/gf/time_invariant_keldysh.jl:
```diff
@@ -0,0 +1,88 @@
+  grid::KeldyshTimeGrid
+  gtr::AntiHermitianToeplitzStorage{T,scalar}
+  les::AntiHermitianToeplitzStorage{T,scalar}
+  ξ::GFSignEnum
```
Same question as with `ξ` in `KeldyshTimeGF`.

krivenko commented on src/Keldysh.jl:
```diff
@@ -1,36 +1,36 @@
-
-# generate_gf functions
-export gf_1level, dos2gf
+export DeltaDOS, flat_dos, gaussian_dos, bethe_dos, chain_dos, square_dos
```
`AbstractDOS`, `DOS` and `SingularDOS`?

krivenko commented on src/alps.jl:
```diff
@@ -0,0 +1,153 @@
+struct ALPSTimeGF
+  G
+  function ALPSTimeGF(G::T) where T <: AbstractTimeGF
+    @assert norbitals(G) == 1
```
Correct me if I am wrong, but I'd make this check static by writing
```julia
function ALPSTimeGF(G::T) where T <: AbstractTimeGF{ComplexF64,true}
```

krivenko commented on test/time_gf.jl:
```diff
@@ -0,0 +1,87 @@
+      end
+    end
+
+    grid_fine = FullTimeGrid(c, 41, 101)
```
Unused?

krivenko commented on src/nessi.jl:
```diff
@@ -47,15 +58,75 @@ function nessi_read_tv(h5g::HDF5.Group)
-  tv  = nessi_read_tv(h5g)
-  mat = nessi_read_mat(h5g)
-  return (;les, ret, tv, mat)
+function NessiGFData(h5g::HDF5.Group)
```
@HugoStrand Hugo, would you mind checking this function? You are apparently the best person for this task.

krivenko commented on examples/anderson_nca.jl:
```diff
@@ -320,7 +292,7 @@ function main()
     h5f["/output/obs/pop/rho"] = ρt
     h5f["/output/obs/pop/Z"] = Zt
-    h5f["/output/obs/pop/t"] = t
+    h5f["/output/obs/pop/t"] = collect(t)
```
`t` is already collected.

krivenko reviewed (changes requested):
In addition to the comments, I would like to remind about the feature request we discussed a few days back. We need `getitem()`/`setitem()` methods that take a pair of time grid points and a pair of GF matrix indices.

kleinhenz committed:
[misc] add compat entry for HDF5

kleinhenz committed:
[time_grid] cleanup

kleinhenz committed:
[dos] export more types

kleinhenz committed:
[hdf5] restrict type signature of ALPSTimeGF constructor

kleinhenz committed:
[time_gf] fix interp test

kleinhenz committed:
[nca] minor cleanup

kleinhenz committed:
[hdf5] fix bug in loading ALPSTimeGF

kleinhenz committed:
[storage] remove unnecessary method

kleinhenz committed:
[time_grid] add function to get number of points on a branch

kleinhenz committed:
[time_grid] getindex method

  Replace grid[branch, i] with grid[branch][i].
  With previous method definition grid[branch, end] does not behave
  correctly

kleinhenz committed:
[time_grid] val -> bpoint

kleinhenz committed:
[test] make tests independently runable

kleinhenz committed:
[storage] cleanup

* CirculantStorage -> PeriodicStorage
* add getindex methods for accessing individual elements
* add tests

kleinhenz committed:
[storage] define some methods for AbstractStorage to avoid duplication

kleinhenz committed:
[storage] further reduce duplication

kleinhenz committed:
[time_gf] allow accessing individual elements

kleinhenz committed:
[time_gf] add is_scalar function + improve comments/docs

kleinhenz committed:
[time_gf] fix method ambiguity

kleinhenz committed:
[time_grid] idx -> cidx

kleinhenz committed:
[doc] add example to README

kleinhenz committed:
[time_gf] add TODOs

kleinhenz committed:
[time_gf] minor cleanup

kleinhenz committed:
[misc] add *.h5 to .gitignore

kleinhenz commented:
ok, I think I have addressed everything. Going to merge and tag a new release.
